### PR TITLE
Load lines and modes from TransXChange

### PR DESCRIPTION
**Pull request by woshilapin**

### CODE REVIEW

File: src/transxchange/read.rs [hunk 1]
```diff
@@ -177,6 +316,14 @@ fn read_xml(transxchange: &Element, collections: &mut Collections) -> Result<()>
     }
     let companies = load_companies(transxchange)?;
     collections.companies.merge(companies);
+    let (commercial_mode, physical_mode) = load_commercial_physical_modes(transxchange)?;
```
Comment (ArnaudOggy) [inline src/transxchange/read.rs]: Warn and continue if mode absent in xml or not in MODES list.

Comment (woshilapin) [reply src/transxchange/read.rs]: I discussed with @papailio and possibly, we're gonna end up with a default mode when it's not present or incorrect. That said, the default value has not been defined yet. Therefore, I did refactor the code to make it easy to ignore completely one file at any point of the parsing (in this case, if `load_commercial_physical_modes()` returns an error.

File: src/transxchange/read.rs [hunk 2]
```diff
@@ -168,15 +186,105 @@ fn load_companies(transxchange: &Element) -> Result<CollectionWithId<Company>> {
...
+) -> Result<(CommercialMode, PhysicalMode)> {
+    let raw_mode = transxchange
+        .try_only_child("Services")?
+        .try_only_child("Service")?
+        .try_only_child("Mode")?
+        .text();
+    let mode = MODES
+        .get(raw_mode.as_str())
+        .ok_or_else(|| {
+            format_err!(
```
Comment (ArnaudOggy) [inline src/transxchange/read.rs]: As seen together apply the default bus mode

File: src/transxchange/read.rs [hunk 3]
```diff
@@ -168,15 +186,105 @@ fn load_companies(transxchange: &Element) -> Result<CollectionWithId<Company>> {
...
         update_validity_period_from_transxchange(&mut collections.datasets, transxchange)?;
     let network = load_network(transxchange)?;
-    if collections.networks.get(&network.id).is_none() {
-        collections.networks.push(network)?;
-    }
     let companies = load_companies(transxchange)?;
+    let (commercial_mode, physical_mode) = match load_commercial_physical_modes(transxchange) {
+        Ok(modes) => modes,
+        Err(e) => {
+            warn!("Ignore the TransXChange file: {}", e);
```
Comment (ArnaudOggy) [inline src/transxchange/read.rs]: Now useless with the application of a default mode :confused:

File: src/transxchange/read.rs [hunk 4]
```diff
@@ -168,15 +186,105 @@ fn load_companies(transxchange: &Element) -> Result<CollectionWithId<Company>> {
...
+
+fn load_line(transxchange: &Element, network_id: &str, commercial_mode_id: &str) -> Result<Line> {
+    let service = transxchange
+        .try_only_child("Services")?
+        .try_only_child("Service")?;
+    let id = service.try_only_child("ServiceCode")?.text();
+    let code = Some(
+        service
+            .try_only_child("Lines")?
+            .try_only_child("Line")?
```
Comment (ArnaudOggy) [inline src/transxchange/read.rs]: In rare cases there are several lines, so get all children
See /W/NewMapsFirstSomerset&Avon7XP-Newport-Bristol5_TXC_2019802-1151_FSAO007X.xml

File: src/transxchange/read.rs [hunk 5]
```diff
@@ -168,15 +186,105 @@ fn load_companies(transxchange: &Element) -> Result<CollectionWithId<Company>> {
...
+        ..Default::default()
+    };
+    Ok((commercial_mode, physical_mode))
+}
+
+fn load_line(transxchange: &Element, network_id: &str, commercial_mode_id: &str) -> Result<Line> {
+    let service = transxchange
+        .try_only_child("Services")?
+        .try_only_child("Service")?;
+    let id = service.try_only_child("ServiceCode")?.text();
```
Comment (ArnaudOggy) [inline src/transxchange/read.rs]: Concatenate ServiceCode and Lineid (several lines possible with the same ServiceCode)

**Comment by woshilapin** [comment]
I introduced a default mode for files that:
- don't have a `<Mode>`
- have an unknown mode

File: src/transxchange/read.rs (hunk 1 above)
Comment (woshilapin) [reply src/transxchange/read.rs]: Done in 9837c714c2044c279627152b80c554924f5198d8.

File: src/transxchange/read.rs (hunk 2 above)
Comment (woshilapin) [reply src/transxchange/read.rs]: Done in 9837c714c2044c279627152b80c554924f5198d8.

File: src/transxchange/read.rs (hunk 3 above)
Comment (woshilapin) [reply src/transxchange/read.rs]: Changed indeed in 9837c714c2044c279627152b80c554924f5198d8.

File: src/transxchange/read.rs (hunk 5 above)
Comment (woshilapin) [reply src/transxchange/read.rs]: Done in 4990a7ad2e6561f105f021feb354944130648383.

File: src/transxchange/read.rs (hunk 4 above)
Comment (woshilapin) [reply src/transxchange/read.rs]: Done in 4990a7ad2e6561f105f021feb354944130648383.

**Comment by ArnaudOggy** [comment]
Add control on file_content.parse() because xml file can be empty (like suf_9-OC7-_-y08-1.xml).
Skip the file in this case (and warn).

**Comment by woshilapin** [comment]
I added a security about files that cannot be parsed as XML files. See b87dc0de13a82cd0138823dd17806340924f1c5a.

File: src/transxchange/read.rs [hunk 6]
```diff
@@ -168,15 +187,100 @@ fn load_companies(transxchange: &Element) -> Result<CollectionWithId<Company>> {
...
+) -> Result<(CommercialMode, PhysicalMode)> {
+    let mode = match transxchange
+        .try_only_child("Services")?
+        .try_only_child("Service")?
+        .try_only_child("Mode")
+        .map(Element::text)
+    {
+        Ok(mode) => MODES.get(mode.as_str()).unwrap_or(&DEFAULT_MODE),
+        Err(e) => {
+            warn!("{}", e);
```
Comment (ArnaudOggy) [inline src/transxchange/read.rs]: proposal
`warn!("{} - Default mode '{}' assigned", e, DEFAULT_MODE);`

File: src/transxchange/read.rs [hunk 7]
```diff
@@ -168,15 +187,100 @@ fn load_companies(transxchange: &Element) -> Result<CollectionWithId<Company>> {
...
+    let service = transxchange
+        .try_only_child("Services")?
+        .try_only_child("Service")?;
+    let service_id = service.try_only_child("ServiceCode")?.text();
+    let mut lines = CollectionWithId::default();
+    for line in service.try_only_child("Lines")?.children() {
+        if let Some(line_id) = line.attr("id") {
+            let id = format!("{}:{}", service_id, line_id);
+            let code = Some(line.try_only_child("LineName")?.text().trim().to_string());
+            let name = if let Ok(description) = service.try_only_child("Description") {
```
Comment (ArnaudOggy) [inline src/transxchange/read.rs]: name, forward_name, backward_name, network_id and commercial_mode_id 
can be out of the loop (if you want)

File: src/transxchange/read.rs (hunk 6 above)
Comment (woshilapin) [reply src/transxchange/read.rs]: I like it better indeed. Done in 50d910bd02903b8a4123c6fd20b9cc5e5044227f.

File: src/transxchange/read.rs (hunk 7 above)
Comment (woshilapin) [reply src/transxchange/read.rs]: Good point. Done in 2b710d24a4b4bea5ae6fc692c331abb5b37db960.

**Pull request merged**